src/chainrules.jl

```diff
@@ -0,0 +1,142 @@
+## Reverse Rules Delta
+
+function rrule(::typeof(Distances.evaluate), s::Delta, x::AbstractVector, y::AbstractVector)
+    d = evaluate(s, x, y)
+    function evaluate_pullback(::Any)
+        return NO_FIELDS, Zero(), Zero()
+    end
+    return d, evaluate_pullback
+end
+
+function rrule(
+    ::typeof(Distances.pairwise), d::Delta, X::AbstractMatrix, Y::AbstractMatrix; dims=2
+)
+    P = Distances.pairwise(d, X, Y; dims=dims)
+    function pairwise_pullback(::Any)
+        return NO_FIELDS, Zero(), Zero()
+    end
+    return P, pairwise_pullback
+end
+
+function rrule(::typeof(Distances.pairwise), d::Delta, X::AbstractMatrix; dims=2)
+    P = Distances.pairwise(d, X; dims=dims)
+    function pairwise_pullback(::Any)
+        return NO_FIELDS, Zero()
+    end
+    return P, pairwise_pullback
+end
+
+function rrule(::typeof(Distances.colwise), d::Delta, X::AbstractMatrix, Y::AbstractMatrix)
+    C = Distances.colwise(d, X, Y)
+    function colwise_pullback(::AbstractVector)
+        return NO_FIELDS, Zero(), Zero()
+    end
+    return C, colwise_pullback
+end
+
+## Reverse Rules DotProduct
+function rrule(
+    ::typeof(Distances.evaluate), s::DotProduct, x::AbstractVector, y::AbstractVector
+)
+    d = dot(x, y)
+    function evaluate_pullback(Δ)
+        return NO_FIELDS, Δ .* y, Δ .* x
+    end
+    return d, evaluate_pullback
+end
+
+function rrule(
+    ::typeof(Distances.pairwise),
+    d::DotProduct,
+    X::AbstractMatrix,
+    Y::AbstractMatrix;
+    dims=2,
+)
+    P = Distances.pairwise(d, X, Y; dims=dims)
+    if dims == 1
+        function pairwise_pullback_cols(Δ)
+            return NO_FIELDS, Δ * Y, Δ' * X
+        end
+        return P, pairwise_pullback_cols
+    else
+        function pairwise_pullback_rows(Δ)
+            return NO_FIELDS, Y * Δ', X * Δ
+        end
+        return P, pairwise_pullback_rows
+    end
+end
+
+function rrule(::typeof(Distances.pairwise), d::DotProduct, X::AbstractMatrix; dims=2)
+    P = Distances.pairwise(d, X; dims=dims)
+    if dims == 1
+        function pairwise_pullback_cols(Δ)
+            return NO_FIELDS, 2 * Δ * X
+        end
+        return P, pairwise_pullback_cols
+    else
+        function pairwise_pullback_rows(Δ)
+            return NO_FIELDS, 2 * X * Δ
+        end
+        return P, pairwise_pullback_rows
+    end
+end
+
+function rrule(
+    ::typeof(Distances.colwise), d::DotProduct, X::AbstractMatrix, Y::AbstractMatrix
+)
+    C = Distances.colwise(d, X, Y)
+    function colwise_pullback(Δ::AbstractVector)
+        return (nothing, Δ' .* Y, Δ' .* X)
+    end
+    return C, colwise_pullback
+end
+
+## Reverse Rules Sinus
+function rrule(::typeof(Distances.evaluate), s::Sinus, x::AbstractVector, y::AbstractVector)
+    d = (x - y)
+    sind = sinpi.(d)
+    val = sum(abs2, sind ./ s.r)
+    gradx = 2π .* cospi.(d) .* sind ./ (s.r .^ 2)
```

Comment:
Do we depend on StatsFuns? If yes, one could also use
```suggestion
    gradx = twoπ .* cospi.(d) .* sind ./ (s.r .^ 2)
```